hex-literal/src/lib.rs

```diff
@@ -1,169 +1,86 @@
-    /// # Panics
-    /// This panics if a non-hex, non-whitespace character is encountered.
-    fn next_hex_val(&mut self) -> Option<u8> {
-        loop {
-            let v = self.buf.next()?;
-            let n = match v {
-                b'0'..=b'9' => v - 48,
-                b'A'..=b'F' => v - 55,
-                b'a'..=b'f' => v - 87,
-                b' ' | b'\r' | b'\n' | b'\t' => continue,
-                0..=127 => panic!("encountered invalid character: `{}`", v as char),
-                _ => panic!("encountered invalid non-ASCII character"),
-            };
-            return Some(n);
+/// Decode hex strings into a byte array of pre-computed length.
+///
+/// This function is an implementation detail and SHOULD NOT be called directly!
+#[doc(hidden)]
+pub const fn decode<const LEN: usize>(strings: &[&[u8]]) -> [u8; LEN] {
+    let mut i = 0;
+    let mut buf = [0u8; LEN];
+    let mut buf_pos = 0;
+    while i < strings.len() {
+        let mut pos = 0;
+        while let Some((byte, new_pos)) = next_byte(strings[i], pos) {
+            buf[buf_pos] = byte;
+            buf_pos += 1;
+            pos = new_pos;
         }
+        i += 1;
     }
-}
-
-impl Iterator for TokenTreeIter {
-    type Item = TokenTree;
-
-    /// Produces hex values (as `u8` literals) parsed from the `TokenTreeIter`'s
-    /// internal buffer, alternating with commas to separate the elements of the
-    /// generated array of bytes.
-    ///
-    /// # Panics
-    /// This panics if the internal buffer contains an odd number of hex
-    /// characters.
-    fn next(&mut self) -> Option<TokenTree> {
-        let v = if self.is_punct {
-            TokenTree::Punct(Punct::new(',', Spacing::Alone))
-        } else {
-            let p1 = self.next_hex_val()?;
-            let p2 = match self.next_hex_val() {
-                Some(v) => v,
-                None => panic!("expected even number of hex characters"),
-            };
-            let val = (p1 << 4) + p2;
-            TokenTree::Literal(Literal::u8_suffixed(val))
-        };
-        self.is_punct = !self.is_punct;
-        Some(v)
+    if LEN != buf_pos {
+        panic!("Length mismatch. Please report this bug.");
     }
+    buf
 }
 
 /// Macro for converting sequence of string literals containing hex-encoded data
 /// into an array of bytes.
-#[proc_macro]
-pub fn hex(input: TokenStream) -> TokenStream {
-    let mut out_ts = TokenStream::new();
-    for tt in ignore_groups(input) {
-        let iter = match tt {
-            TokenTree::Literal(literal) => TokenTreeIter::new(literal),
-            unexpected => panic!("expected string literal, got `{}`", unexpected),
-        };
-        out_ts.extend(iter);
-    }
-    TokenStream::from(TokenTree::Group(Group::new(Delimiter::Bracket, out_ts)))
+#[macro_export]
+macro_rules! hex {
+    ($($s:literal)*) => {{
+        const STRINGS: &[&'static [u8]] = &[$($s.as_bytes(),)*];
+        $crate::decode::<{ $crate::len(STRINGS) }>(STRINGS)
```

Comment:
This line may not be const-promoted (or const-evaluated), which may causes performance regression.